main.go

```diff
@@ -121,64 +79,64 @@ func main() {
 }
 
 func tsproxy(ctx context.Context) error {
-	var (
-		state = flag.String("state", "", "Optional directory for storing Tailscale state.")
-		tslog = flag.Bool("tslog", false, "If true, log Tailscale output.")
-		port  = flag.Int("port", 32019, "HTTP port for metrics and service discovery.")
-	)
-	var upstreams upstreamFlag
-	flag.Var(&upstreams, "upstream", "Repeated for each upstream. Format: name=http://backend:8000")
-	flag.Parse()
+	if len(os.Args) != 2 {
+		return fmt.Errorf("usage: %s <path to config>", os.Args[0])
+	}
+	cfgb, err := os.ReadFile(os.Args[1])
+	if err != nil {
+		return fmt.Errorf("reading config file %s: %w", os.Args[1], err)
+	}
+	cfg, err := parseAndValidateConfig(cfgb)
+	if err != nil {
+		return fmt.Errorf("reading config file %s: %w", os.Args[1], err)
+	}
 
-	if len(upstreams) == 0 {
+	if len(cfg.Upstreams) == 0 {
 		return fmt.Errorf("required flag missing: upstream")
 	}
-	if *state == "" {
-		v, err := os.UserCacheDir()
-		if err != nil {
-			return err
-		}
-		dir := filepath.Join(v, "tsproxy")
-		if err := os.MkdirAll(dir, 0o700); err != nil {
-			return err
-		}
-		state = &dir
-	}
 
 	logger := slog.New(slog.NewJSONHandler(os.Stderr, &slog.HandlerOptions{}))
 	slog.SetDefault(logger)
 
 	// If tailscaled isn't ready yet, just crash.
-	st, err := (&tailscale.LocalClient{}).Status(ctx)
-	if err != nil {
-		return fmt.Errorf("tailscale: get node status: %w", err)
-	}
-	if v := len(st.Self.TailscaleIPs); v != 2 {
-		return fmt.Errorf("want 2 tailscale IPs, got %d", v)
-	}
+	/*
+		st, err := (&tailscale.LocalClient{}).Status(ctx)
+		if err != nil {
+			return fmt.Errorf("tailscale: get node status: %w", err)
+		}
+		if v := len(st.Self.TailscaleIPs); v != 2 {
+			return fmt.Errorf("want 2 tailscale IPs, got %d", v)
+		}
+	*/
 
 	// service discovery targets (self + all upstreams)
-	targets := make([]target, len(upstreams)+1)
+	targets := make([]target, len(cfg.Upstreams)+1)
 
 	var g run.Group
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 	g.Add(run.SignalHandler(ctx, os.Interrupt, syscall.SIGTERM))
 
 	{
-		p := strconv.Itoa(*port)
+		p := strconv.Itoa(cfg.MetricsDiscoveryPort)
 
 		var listeners []net.Listener
-		for _, ip := range st.Self.TailscaleIPs {
-			ln, err := net.Listen("tcp", net.JoinHostPort(ip.String(), p))
-			if err != nil {
-				return fmt.Errorf("listen on %s:%d: %w", ip, *port, err)
-			}
-			listeners = append(listeners, ln)
+		/*
+			for _, ip := range st.Self.TailscaleIPs {
+				ln, err := net.Listen("tcp", net.JoinHostPort(ip.String(), p))
+				if err != nil {
+					return fmt.Errorf("listen on %s:%d: %w", ip, cfg.MetricsDiscoveryPort, err)
+				}
+				listeners = append(listeners, ln)
+			}*/
+		ln, err := net.Listen("tcp", "0.0.0.0:"+p)
```

Comment:
I use `http://nuc10i5/sd` in my prom config, I guess I could make tsproxy.service listen on the node's tailscale IP using some ansible: `{{ ansible_facts.tailscale0.ipv4.address }`

alternatively could you check for the presence of a local socket? will think about it too.